gazebo_ros/src/executor.cpp

```diff
@@ -34,14 +49,100 @@ Executor::~Executor()
   spin_thread_.join();
 }
 
-void Executor::run()
+void Executor::shutdown()
 {
-  spin();
+  rclcpp::shutdown();
 }
 
-void Executor::shutdown()
+void
+Executor::add_node(
+  rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_ptr, bool notify)
 {
-  rclcpp::shutdown();
+  std::lock_guard<std::mutex> wait_lock(wait_mutex_);
+  rclcpp::Executor::add_node(node_ptr, notify);
+}
+
+void
+Executor::add_node(std::shared_ptr<rclcpp::Node> node_ptr, bool notify)
+{
+  this->add_node(node_ptr->get_node_base_interface(), notify);
+}
+
+void
+Executor::remove_node(
+  rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_ptr, bool notify)
+{
+  std::lock_guard<std::mutex> wait_lock(wait_mutex_);
+  rclcpp::Executor::remove_node(node_ptr, notify);
+}
+
+void
+Executor::remove_node(std::shared_ptr<rclcpp::Node> node_ptr, bool notify)
+{
+  this->remove_node(node_ptr->get_node_base_interface(), notify);
+}
+
+void
+Executor::spin()
+{
+  if (spinning.exchange(true)) {
+    throw std::runtime_error("spin() called while already spinning");
+  }
+  RCLCPP_SCOPE_EXIT(this->spinning.store(false); );
+  std::vector<std::thread> threads;
+  size_t thread_id = 0;
+  {
+    for (; thread_id < number_of_threads_ - 1; ++thread_id) {
+      auto func = std::bind(&Executor::run, this);
+      threads.emplace_back(func);
+    }
+  }
+
+  run();
+  for (auto & thread : threads) {
+    thread.join();
+  }
+}
+
+void
+Executor::run()
+{
+  while (rclcpp::ok(this->context_) && spinning.load()) {
+    rclcpp::AnyExecutable any_exec;
+    std::lock_guard<std::mutex> wait_lock(wait_mutex_);
+    if (!rclcpp::ok(this->context_) || !spinning.load()) {
+      return;
+    }
+
+    if (!get_next_executable(any_exec, next_exec_timeout_)) {
```

Comment:
why the timeout is needed?

I see you mentioned a deadlock in the PR post, but I'm not sure of how that's happening.

Reply:
I don't recall exactly, but I'mp pretty sure I was running into deadlock when adding/removing nodes from separate threads. I think with the changes in f9c0b54, a timeout is no longer required.